File: python/problem5.py
#!/usr/bin/env python2.5

"""
Problem 5 from Project Euler

What is the smallest number that is evenly divisible by all of the numbers from 1 to 20?
"""


#List of numbers
#If another number is a factor of it that number, remove it from the list
#Multiply the numbers that are left

from math import sqrt

n = 20
ldl = range(1, n+1)


def factor(n):
    f = []
    i = 1

    while True:
        if i > sqrt(n): break
        if n % i == 0: f.append(i)
        i += 1

    return f

def test(n, total):
    divall = False
    divs = range(1, n+1)
    for i in divs:
        if not total % i == 0:
            divall = False
            break
        else:
            divall = True

    return divall


#for lf in ldl:
    #for slf in factor(lf)[:-1]:
        #for lf2 in ldl[:ldl.index(lf)]:
            #for slf2 in factor(lf2)[1:]:
                #if slf == slf2:
                    #print "Yep", lf, slf, lf2, slf2
                    #ldl.pop(ldl.index(lf2))

#print ldl

#total = 1
#for i in ldl:
    #total *= i

#print total


total = 1
for t in ldl:
    total *= t
print(total)

for i in range(n, 0, -1):
    if test(n, (total/i)):
        total = total/i

print(total)

File: elixir/problem5.exs
#!/usr/bin/env elixir

# Problem 5:
# 2520 is the smallest number that can be divided by each of the numbers from
# 1 to 10 without any remainder.
# What is the smallest positive number that is evenly divisible by all of the
# numbers from 1 to 20?

Code.load_file("factor.ex")

n = 20

defmodule Euler5 do
  def smallest_common_multiple(n) do
    n..1
    |> Enum.reduce(&Euler5.product/2)
    |> Euler5.divide_all_factors(n, n)
  end

  def divide_all_factors(total, 1, _) do
    total
  end
  def divide_all_factors(total, n, under) do
    test_total = div(total, n)

    if div_under(test_total, under) do
      divide_all_factors(test_total, n-1, under)
    else
      divide_all_factors(total, n-1, under)
    end
  end

  def div_under(_, 1),                             do: true
  def div_under(n, under) when rem(n, under) == 0, do: div_under(n, under-1)
  def div_under(n, under) when rem(n, under) > 0,  do: false

  def product(n, acc) do
    n * acc
  end
end

IO.inspect(Euler5.smallest_common_multiple(n))
